odoo/models.py

```diff
@@ -2010,16 +2007,68 @@ def _read_group_fill_temporal(self, data, groupby, aggregated_fields, annotated_
         if first_a_gby['type'] not in ('date', 'datetime'):
             return data
         interval = first_a_gby['interval']
+        granularity = first_a_gby['granularity']
         groupby_name = groupby[0]
 
         # existing non null datetimes
-        existing = [d[groupby_name] for d in data if d[groupby_name]]
-
-        if len(existing) < 2:
+        existing = sorted([d[groupby_name] for d in data if d[groupby_name]])
+        existing_from = existing[0] if existing else False
+        existing_to = existing[-1] if existing else False
+
+        # fill temporal between bounds
+        # the context keys : fill_temporal_from and fill_temporal_to offer a way to guarantee
+        # that at least a certain date range will be returned as groups. These are also the
+        # bounds to fill. Any group outside those bounds will not be lost, but there will be
+        # no empty group added. If any of those keys are omitted, existing bounds are used, if
+        # applicable. If fill_temporal_from is not specified, and there is no data, no group
+        # are returned.
+        # Example:
+        # existing:           [Jan, Sep]
+        # fill_temporal_from: [Apr]
+        # fill_temporal_to:   [Jul]
+        # result:             [Jan, |Apr, May, Jun, Jul|, Sep]
+        tz = next((date.tzinfo for date in existing if date.tzinfo), False)
+        fill_from = self.env.context.get('fill_temporal_from')
+        fill_to = self.env.context.get('fill_temporal_to')
+        if fill_from:
+            fill_from = date_utils.start_of(datetime.datetime.strptime(fill_from, "%Y-%m-%d"), granularity)
+            if tz:
+                fill_from = tz.localize(fill_from)
+        elif existing_from:
+            fill_from = existing_from
+        else:
             return data
-
-        # assumption: existing data is sorted by field 'groupby_name'
-        first, last = existing[0], existing[-1]
+        if fill_to:
+            fill_to = date_utils.end_of(datetime.datetime.strptime(fill_to, "%Y-%m-%d"), granularity)
+            if tz:
+                fill_to = tz.localize(fill_to)
+        elif existing_to:
+            fill_to = existing_to
+        else:
+            fill_to = fill_from
+
+        if fill_to and fill_to < fill_from:
+            raise ValueError("Invalid context keys : from %s to %s" % (str(fill_from), str(fill_to)))
+
+        required = list(date_utils.date_range(fill_from, fill_to, interval))
```

Comment:
required what ? `required_groups` I guess.
The naming looks more like you have a bool than a list of month/day/etc..